Add pre-commit hooks for code formatting and linting

iswariyam opened the pull request:
Pre-commit hooks have been added to format code and check for linting errors before commit changes in git locally.

## Changelog
* Add .precommit_hooks.yaml
* Add .isort.cfg to sort imports
* Other files in the repo formatted by the pre-commit hooks

## Related PRs

Closes #147 
Related to #148 

## Checklist:

- [x] My code doesn't contain unnecessary comment blocks (e.g. unused code, templates of `package.xml` or `CMakeLists.txt`)
- [x] I have updated the `package.xml` and `CMakeLists.txt` with the correct dependencies.
- [x] I have updated the documentation accordingly.

iswariyam committed:
Add pre-commit hooks

iswariyam committed:
Add cpp linting to precommit hooks

iswariyam committed:
Disable pylint and cpplint

iswariyam committed:
Update README.md

iswariyam committed:
Fix wrongly formatted rosinstall file

sthoduka commented:
for the .cpp files, it has changed the indentation from 4 spaces to 2 spaces. Is this intentional?

sthoduka commented on README.md:
```diff
@@ -1,152 +1,158 @@
 If you do not have a Ubuntu distribution on your computer you can download it here
 
-     http://www.ubuntu.com/download
+     http: // www.ubuntu.com / download
```
why are URLs edited like this?

iswariyam replied:
fixed. This is one of those files which got messed up when I was trying out different hooks.

sthoduka commented on docs/source/conf.py:
```diff
@@ -12,25 +12,26 @@
+
 def get_git_short_hash():
-    rc = subprocess.check_output(['git', 'rev-parse', '--short', 'HEAD'])
+    rc = subprocess.check_output(["git", "rev-parse", "--short", "HEAD"])
```
is using double quotes the recommended way?

iswariyam replied:
The python code formater used here is Black. As per black code style double quotes in preferred (https://black.readthedocs.io/en/stable/the_black_code_style.html#strings).

sthoduka commented on mir_calibration/mir_kinect_calibration/common/src/transformation_estimation.cpp:
```diff
@@ -1,42 +1,40 @@
+  * with x, y, z, roll, pitch, and yaw of the estimated transform. */
+void estimateTransformation(float *a, float *b, int point_count,
+                            float *transform) {
+  pcl::PointCloud<pcl::PointXYZ> A, B;
```
this is an example where it's just the indentation length that has changed

iswariyam replied:
Yes, this is because clang-format google style is used in the pre-commit hooks. As per google style guide, the code is indented 2 spaces at a time (https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs).

iswariyam committed:
Fix urls in readme

iswariyam committed:
Enable linting

sthoduka commented:
@iswariyam before we merge this, should #148 and #150 be solved first? Or can I merge it now (after fixing the conflicts)?

iswariyam commented:
> @iswariyam before we merge this, should #148 and #150 be solved first? Or can I merge it now (after fixing the conflicts)?

#148 and #150 can be fixed later. This PR can merged without #148 and #150 being resolved.

sthoduka commented:
is there a way to change the following? I couldn't figure out where they are set.
1. the 80 character per line limit (to 120 for example)
2. move the curly brace to a new line
I know both of those are a matter of preference, but I think the way it is now makes the code less readable. I was comparing [this](https://github.com/iswariyam/mas_industrial_robotics/blob/feature/precommit_hooks/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_ros.cpp) and [this](https://github.com/b-it-bots/mas_industrial_robotics/blob/kinetic/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_node.cpp) for example. It's harder to tell where the function begins in the first one (for me). Also [here](https://github.com/iswariyam/mas_industrial_robotics/blob/feature/precommit_hooks/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_node.cpp) the 80 character limit just adds lines.

I've fixed the merge conflicts, but waiting to merge it if it's easy to change those two things.

iswariyam commented:
> for the .cpp files, it has changed the indentation from 4 spaces to 2 spaces. Is this intentional?



> is there a way to change the following? I couldn't figure out where they are set.
> 
>     1. the 80 character per line limit (to 120 for example)
> 
>     2. move the curly brace to a new line
>        I know both of those are a matter of preference, but I think the way it is now makes the code less readable. I was comparing [this](https://github.com/iswariyam/mas_industrial_robotics/blob/feature/precommit_hooks/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_ros.cpp) and [this](https://github.com/b-it-bots/mas_industrial_robotics/blob/kinetic/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_node.cpp) for example. It's harder to tell where the function begins in the first one (for me). Also [here](https://github.com/iswariyam/mas_industrial_robotics/blob/feature/precommit_hooks/mir_perception/mir_object_segmentation/ros/src/scene_segmentation_node.cpp) the 80 character limit just adds lines.
> 
> 
> I've fixed the merge conflicts, but waiting to merge it if it's easy to change those two things.

Yes, they can be modified. As of now the clang-format configurable style options take on default google style values. I could add a .clang-format file where the desired options could be set. This is explained [here](https://releases.llvm.org/3.7.1/tools/docs/ClangFormatStyleOptions.html) (under Configurable format style options) in detail.
1. BreakBeforeBraces option (as explained in the above link) can be set to Allman so that the braces occur in a new line in a func/class/for loop/if else statement.
2. Are you sure you want the column width/line length to be set to 120? Even the llvm style uses only 100 as the column width.

Please have a look at the other format style options and let me know if anything should be customized.

sthoduka commented:
ah ok, ya a separate .clang-format file would be good. 
yes, a column width of 100 is fine too. I think for now those are the only two settings.

iswariyam committed:
Add .clang-format for cpp file formatting
.clang-format uses google style
.clang-format uses Linux option for breakbeforebraces
existing cpp files are formatted
clang-format used from diff precommit hooks repo to parse .clang-format during commits
disabling cpplint to commit changes

iswariyam committed:
Merge from kinetic branch

iswariyam committed:
Enable cpp lint

iswariyam commented:
I have added .clang-format file and hence clang has again formatted the cpp files in the repo according to the rules in .clang-format. Also, I resolved the merge conflicts. As a result I had to fix the python linting errors in planner_wrapper.py such as renaming constants, removing except statements with pass. .pylintrc has also been added to customize some configurations.

iswariyam commented on .pylintrc:
```diff
@@ -0,0 +1,553 @@
+        deprecated-sys-function,
+        exception-escape,
+        comprehension-escape,
+        R0205,  # Python 2.7 classes require object
```
Here is where I have disabled some of the rules because we have code written in python 2.7 but pre-commit hooks only support python >=3.5. So essentially all checks are run in a python 3.7 environment as per pre-commit hooks config.yaml file.

iswariyam commented on .clang-format:
```diff
@@ -0,0 +1,90 @@
+  BeforeElse:      false
+  IndentBraces:    false
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Linux
```
This option is set to Linux instead of Allman because Allman breaks after if/else which clashes with cppcheck in pre-commit hooks config.yaml. Linux breaks braces after classes, functions and namespaces.

sthoduka commented:
thanks!
